### PR TITLE
doc: Make PR description a contributor expectation

### PR DESCRIPTION
**Pull request by decsny**
Make a PR description an expectation/requirement for pull requests submitted to the zephyr project.

As a reviewer, I am getting annoyed with how many PRs that show up that are non trivial with no description. It does not seem very respectful to reviewers when a contributor neglects to put even a single character in the description, and this makes me not want to review the PR. Descriptions are helpful to have an idea of what I am about to review rather than starting by piecing together context clues.

### CODE REVIEW

File: doc/contribute/contributor_expectations.rst [hunk 1]
```diff
@@ -93,6 +93,9 @@ PR Requirements
 - Each commit in the PR must provide a commit message following the
   :ref:`commit-guidelines`.
 
+- The PR must include a description that summarizes the changes in the PR
```
Comment (nashif) [inline doc/contribute/contributor_expectations.rst:96]: nice,  but do you think people will read this? Still good to get in, but IMO we need to enforce this in CI, fail PRs without any description.

**Comment by kartben** [comment]
It might be enough to just set a pull request template? https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository

**Comment by decsny** [comment]
I think why not all 3 of these things, document the expectation, make a pull request template, and fail CI if there is no description

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

File: doc/contribute/contributor_expectations.rst (hunk 1 above)
Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:96]: Don't we do it already?

File: doc/contribute/contributor_expectations.rst [hunk 2]
```diff
@@ -93,6 +93,9 @@ PR Requirements
...
+  by describing what the changes are and why they are being made.
```
Comment (fabiobaltieri) [inline doc/contribute/contributor_expectations.rst:97]: I would swap this, the important part is the "why", the "what" you can tell from the code itself (hopefully). How about "by describing why the change is needed and what the change is"?

Comment (decsny) [reply doc/contribute/contributor_expectations.rst:97]: Personally disagree, that wording sounds awkward to my ears. Also, I think the 'what' is actually the main motivation for this PR, because it is annoying to see a PR with no description and try to figure out from the code what it's doing. The 'why' is only important to help form a judgement about the PR, but the first step is understanding the 'what'

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:97]: Well the "what" should be already covered by the subject. Anyway, both are mentioned (and yeah it would sound awkward) so fine by me either way.

**Review by fabiobaltieri — DISMISSED** [review]

File: doc/contribute/contributor_expectations.rst (hunk 2 above)
Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:97]: It depends on the change, ideally the code is perfectly understandable and does not need a verbal description and the oneliner tells everything you need to know to review the PR. Reality is different, obviously.

Comment (henrikbrixandersen) [reply doc/contribute/contributor_expectations.rst:97]: I disagree. We need more than a one-liner, which is also why we require a commit message _body_.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:97]: Sorry I wasn't trying to say that we need only the oneliner. What I wanted to say is that the oneliner should already describe what the change is, and the body should expand on that adding the motivation for doing the change.

Comment (decsny) [reply doc/contribute/contributor_expectations.rst:97]: Maybe for a small PR like less than 30 lines you can tell from the code, but I am not uncommonly seeing hundreds/thousands line change PRs with no description. 

When writers/presenters want to convey some information, especially in a technical forum, they put an introduction and conclusion. Why? If the body already contains all the information what is the point of the introduction and conclusion. Well, it's because human brains are not computers (both the writers and the readers), and without some more explicit framing of the content, even if said content is in written form, there may be some misunderstanding or things missed about what is being conveyed. In school, I always heard something like "tell them what you are going to tell them, tell them, then tell them what you just told them" from professors speaking of technical communication. I think a similar sentiment should and in fact DOES apply to an open source software PR to a large extent, because it's also effectively about conveying information to an audience.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:97]: Let me step back for a sec:

I am NOT (I repeat: NOT) saying that we should allow PR with empty description. Our compliance check requires a description, if it wasn't requiring a description, I would change it to require a description, if we did not have compliance requiring a description, I would ask contributors to add a description. I love descriptions, hopefully that leaves no ambiguity.

That said, if you feel like you NEED the description to understand the change, maybe the change itself is not good enough. Maybe it should be refactored, maybe it should be split, maybe it's absolutely perfect and it's just intrinsically complex and it needs to be explained.

You may see some very understandable code and not even check the commit message and get it in. Or you may see a very badly written change with a perfectly written description, you would block it asking to refactor the code.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:97]: > I am not uncommonly seeing hundreds/thousands line change PRs with no description

Again the compliance check requires a description, you see this in the project, the compliance check is broken and we need to fix it. So if that's the case, point to an example please.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:97]: Ah, @henrikbrixandersen kindly pointed out to me that this is referring to the PR description and somehow I was thinking about the commit message. So yeah scrape half of what I said -- yes ideally the CI would flag empty PR messages.

File: doc/contribute/contributor_expectations.rst (hunk 1 above)
Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:96]: Nevermind, PR description, not commit message -- yeah I agree ideally the CI would check this.

Comment (fabiobaltieri) [reply doc/contribute/contributor_expectations.rst:96]: https://github.com/zephyrproject-rtos/zephyr/pull/75788

**Review by nashif — DISMISSED** [review]

**Comment by decsny** [comment]
@kartben the PR the block CI over this merged before the documentation about the expectation, can you please check this PR since you're assigned

**Review by aescolar — DISMISSED** [review]

File: doc/contribute/contributor_expectations.rst (hunk 2 above)
Comment (kartben) [inline doc/contribute/contributor_expectations.rst:97]: there is a repetition in the sentence that I think is not ideal (...PR must include a description of the changes that describes the changes ...)

```suggestion
- The PR's description must include a summary of the changes and their rationale.
```

**Review by kartben — CHANGES_REQUESTED** [review]
sorry for the delay, even more so as I'm ending up requesting a change after it's been approved by quite a few folks already

File: doc/contribute/contributor_expectations.rst (hunk 2 above)
Comment (decsny) [reply doc/contribute/contributor_expectations.rst]: updated

**Review by kartben — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]
